### DISCUSSION TITLE
How do I protect my project from backward-incompatible changes in file formats?

### DISCUSSION THREAD

**Discussion by Yura52** [prefix-dev/pixi#1061] [Q&A]
Hi! I see that [pixi is currently in alpha](https://github.com/prefix-dev/pixi?tab=readme-ov-file#status) and backward-incompatible changes in file formats are possible, which is totally understandable. Does this mean that a pixi-based project that I publish today can become broken for future versions of pixi? How do I deal with that as an author of the project?

**Comment by ruben-arts** [comment]
This is correct, we currently have don't have a way to limit pixi's version for the `pixi.toml`.

The `pixi.lock` however contains a version field which we compare with the version we support in the pixi that is being run. If that version is higher then the version of pixi that is being run then we error. There are some heavy tests in our back-end library `rattler` that test if pixi stays backwards compatible with old lockfiles. 

A nice way to test your projects compatibility is to use our GitHub action: https://pixi.sh/latest/advanced/github_actions/. If you don't set the `pixi-version` you'll automatically get the latest.

**Reply by Yura52** [reply]
To be more specific, my project will be an open source implementation of a research paper. Often, modulo bug fixes, such projects are not "maintained" in the usual sense, i.e. they basically remain unchanged forever and lack most of the typical infrastructure around them (such as CI). Instead, they only serve as reference implementations of methods proposed in papers, allow reproducing results reported in papers and allow other researchers to start new projects by forking existing ones.

My understanding is that the "version" key is what enables new pixi versions to read old `pixi.lock` files ([example](https://github.com/prefix-dev/pixi/blob/139d15e29a346e4af84afc7c0d86f74db033c1d0/pixi.lock#L1)). Would it be possible to add a similar mechanism to `pixi.toml`?

**Reply by ruben-arts** [reply]
This is very much a use-case we want to support! So we should take a look at how we can prepare for the future!

**Reply by Yura52** [reply]
Is it then an appropriate topic for opening a "feature request" issue?

**Reply by Yura52** [reply]
Btw, one more possible direction for solving the problem is saving the version of pixi itself in the lockfile.

**Reply by ruben-arts** [reply]
Yes please, feel free to open an issue. I think locking the pixi version would not be a complete solution to making sure the `pixi.toml` can still be read. 

Just an extra idea would be that we allow for direct installation of a lockfile, possibly also conda-lock as we can still read them. Because we install directly from lockfiles underwater anyway. `pixi install --frozen` also has the same behavior but it will fail if there is no `pixi.toml`

**Reply by Yura52** [reply]
If I understand the idea correctly, it means that, in some situations, the convention "pixi.toml is the source of truth" stops being true.

One more quick idea: perhaps, `pixi init` can set a format version in `pixi.toml` file automatically, for example:

```
# pixi.toml

...

[pixi]
format-version = 1
```

This is not super elegant to have an extra thing in a file, though in big files this will not make a big difference.
